### DISCUSSION TITLE
Can trained models run in browsers, just for model inference?

### DISCUSSION THREAD

**Discussion by ShangChien** [deepmodeling/deepmd-kit#3100] [Q&A] [answered]
Can trained models be loaded in Chrome browsers like [the usage of node.js](https://github.com/deepmodeling/deepmd-kit/blob/master/doc/inference/nodejs.md)?

**Comment by njzjz** [comment, marked as answer]
It's not easy.

While node.js can directly bind with C/C++ and execute x86_64 binaries, browsers can only execute WebAssembly (Wasm). While there are successful [examples](https://github.com/emscripten-forge/recipes) for small packages like NumPy, the upstream TensorFlow C++ library is too hard to build. xref: https://github.com/pyodide/pyodide/issues/50

Another resolution is to use tf.js, which does not support all OPs or customized OPs. At least, one needs to write customized OPs in the JS way. Contribution is welcome if anyone is willing to spend time on it.
